Tests failing on Python 3.12

One of the failing tests:

```
________________________________________________________________ TestBuildCLib.test_run ________________________________________________________________

self = <distutils.tests.test_build_clib.TestBuildCLib object at 0x106a09d60>

    @pytest.mark.skipif('platform.system() == "Windows"')
    def test_run(self):
        pkg_dir, dist = self.create_dist()
        cmd = build_clib(dist)
    
        foo_c = os.path.join(pkg_dir, 'foo.c')
        self.write_file(foo_c, 'int main(void) { return 1;}\n')
        cmd.libraries = [('foo', {'sources': [foo_c]})]
    
        build_temp = os.path.join(pkg_dir, 'build')
        os.mkdir(build_temp)
        cmd.build_temp = build_temp
        cmd.build_clib = build_temp
    
        # Before we run the command, we want to make sure
        # all commands are present on the system.
>       ccmd = missing_compiler_executable()

distutils/tests/test_build_clib.py:128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cmd_names = []

    def missing_compiler_executable(cmd_names=[]):
        """Check if the compiler components used to build the interpreter exist.
    
        Check for the existence of the compiler executables whose names are listed
        in 'cmd_names' or all the compiler executables when 'cmd_names' is empty
        and return the first missing executable or None when none is found
        missing.
    
        """
>       from setuptools._distutils import ccompiler, sysconfig, spawn
E       ModuleNotFoundError: No module named 'setuptools'

/opt/homebrew/Cellar/python@3.12/3.12.1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/test/support/__init__.py:1732: ModuleNotFoundError
```

It looks like the reliance on `test.support` has an implicit dependency on Setuptools.

The other class of failure looks like:

```
_______________________________________________________ TestBuildPy.test_byte_compile_optimized ________________________________________________________

self = <distutils.tests.test_build_py.TestBuildPy object at 0x105247620>

    @pytest.mark.skipif('sys.dont_write_bytecode')
    def test_byte_compile_optimized(self):
        project_dir, dist = self.create_dist(py_modules=['boiledeggs'])
        os.chdir(project_dir)
        self.write_file('boiledeggs.py', 'import antigravity')
        cmd = build_py(dist)
        cmd.compile = 0
        cmd.optimize = 1
        cmd.build_lib = 'here'
        cmd.finalize_options()
>       cmd.run()

/Users/jaraco/code/pypa/distutils/distutils/tests/test_build_py.py:115: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/jaraco/code/pypa/distutils/distutils/command/build_py.py:98: in run
    self.byte_compile(self.get_outputs(include_bytecode=0))
/Users/jaraco/code/pypa/distutils/distutils/command/build_py.py:400: in byte_compile
    byte_compile(
/Users/jaraco/code/pypa/distutils/distutils/util.py:462: in byte_compile
    spawn(cmd, dry_run=dry_run)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cmd = '/Users/jaraco/code/pypa/distutils/.tox/py/bin/python', search_path = 1, verbose = 0, dry_run = 0
env = {'COLUMNS': '152', 'HOME': '/Users/jaraco', 'LANG': 'en_US.UTF-8', 'LINES': '53', ...}

    def spawn(cmd, search_path=1, verbose=0, dry_run=0, env=None):  # noqa: C901
        """Run another program, specified as a command list 'cmd', in a new process.
    
        'cmd' is just the argument list for the new process, ie.
        cmd[0] is the program to run and cmd[1:] are the rest of its arguments.
        There is no way to run a program with a name different from that of its
        executable.
    
        If 'search_path' is true (the default), the system's executable
        search path will be used to find the program; otherwise, cmd[0]
        must be the exact path to the executable.  If 'dry_run' is true,
        the command will not actually be run.
    
        Raise DistutilsExecError if running the program fails in any way; just
        return on success.
        """
        # cmd is documented as a list, but just in case some code passes a tuple
        # in, protect our %-formatting code against horrible death
        cmd = list(cmd)
    
        log.info(subprocess.list2cmdline(cmd))
        if dry_run:
            return
    
        if search_path:
            executable = find_executable(cmd[0])
            if executable is not None:
                cmd[0] = executable
    
        env = env if env is not None else dict(os.environ)
    
        if sys.platform == 'darwin':
            from distutils.util import MACOSX_VERSION_VAR, get_macosx_target_ver
    
            macosx_target_ver = get_macosx_target_ver()
            if macosx_target_ver:
                env[MACOSX_VERSION_VAR] = macosx_target_ver
    
        try:
            proc = subprocess.Popen(cmd, env=env)
            proc.wait()
            exitcode = proc.returncode
        except OSError as exc:
            if not DEBUG:
                cmd = cmd[0]
            raise DistutilsExecError(
                "command {!r} failed: {}".format(cmd, exc.args[-1])
            ) from exc
    
        if exitcode:
            if not DEBUG:
                cmd = cmd[0]
>           raise DistutilsExecError(
                "command {!r} failed with exit code {}".format(cmd, exitcode)
            )
E           distutils.errors.DistutilsExecError: command '/Users/jaraco/code/pypa/distutils/.tox/py/bin/python' failed with exit code 1

/Users/jaraco/code/pypa/distutils/distutils/spawn.py:70: DistutilsExecError
----------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------
Traceback (most recent call last):
  File "/var/folders/sx/n5gkrgfx6zd91ymxr2sr9wvw00n8zm/T/tmpznijdwll.py", line 1, in <module>
    from distutils.util import byte_compile
ModuleNotFoundError: No module named 'distutils'
------------------------------------------------------------------ Captured log call -------------------------------------------------------------------
INFO     root:dir_util.py:71 creating here
INFO     root:file_util.py:137 copying boiledeggs.py -> here
INFO     root:util.py:423 writing byte-compilation script '/var/folders/sx/n5gkrgfx6zd91ymxr2sr9wvw00n8zm/T/tmpznijdwll.py'
INFO     root:spawn.py:38 /Users/jaraco/code/pypa/distutils/.tox/py/bin/python /var/folders/sx/n5gkrgfx6zd91ymxr2sr9wvw00n8zm/T/tmpznijdwll.py
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -159,7 +159,8 @@ jobs:
     needs:
     - test
     - collateral
-    - test_cygwin
+    # disabled due to disabled job
+    # - test_cygwin
 
     runs-on: ubuntu-latest
 

--- a/distutils/tests/__init__.py
+++ b/distutils/tests/__init__.py
@@ -7,7 +7,7 @@ distutils.command.tests package, since command identification is done
 by import rather than matching pre-defined names.
 """
 
-def missing_compiler_executable(cmd_names=[]):
+def missing_compiler_executable(cmd_names=[]):  # pragma: no cover
     """Check if the compiler components used to build the interpreter exist.
 
     Check for the existence of the compiler executables whose names are listed

--- a/distutils/tests/__init__.py
+++ b/distutils/tests/__init__.py
@@ -6,3 +6,35 @@ included in distutils.tests as well, instead of using a separate
 distutils.command.tests package, since command identification is done
 by import rather than matching pre-defined names.
 """
+
+def missing_compiler_executable(cmd_names=[]):
+    """Check if the compiler components used to build the interpreter exist.
+
+    Check for the existence of the compiler executables whose names are listed
+    in 'cmd_names' or all the compiler executables when 'cmd_names' is empty
+    and return the first missing executable or None when none is found
+    missing.
+
+    """
+    from distutils import ccompiler, sysconfig, spawn
+    from distutils import errors
+
+    compiler = ccompiler.new_compiler()
+    sysconfig.customize_compiler(compiler)
+    if compiler.compiler_type == "msvc":
+        # MSVC has no executables, so check whether initialization succeeds
+        try:
+            compiler.initialize()
+        except errors.PlatformError:
+            return "msvc"
+    for name in compiler.executables:
+        if cmd_names and name not in cmd_names:
+            continue
+        cmd = getattr(compiler, name)
+        if cmd_names:
+            assert cmd is not None, \
+                    "the '%s' executable is not configured" % name
+        elif not cmd:
+            continue
+        if spawn.find_executable(cmd[0]) is None:
+            return cmd[0]

--- a/distutils/tests/test_build_clib.py
+++ b/distutils/tests/test_build_clib.py
@@ -1,13 +1,11 @@
 """Tests for distutils.command.build_clib."""
 import os
 
-from test.support import missing_compiler_executable
-
 import pytest
 
 from distutils.command.build_clib import build_clib
 from distutils.errors import DistutilsSetupError
-from distutils.tests import support
+from distutils.tests import support, missing_compiler_executable
 
 
 class TestBuildCLib(support.TempdirManager):

--- a/distutils/tests/test_build_ext.py
+++ b/distutils/tests/test_build_ext.py
@@ -16,6 +16,7 @@ import pytest
 from distutils.core import Distribution
 from distutils.command.build_ext import build_ext
 from distutils import sysconfig
+from distutils.tests import missing_compiler_executable
 from distutils.tests.support import (
     TempdirManager,
     copy_xxmodule_c,
@@ -89,7 +90,7 @@ class TestBuildExt(TempdirManager):
         return build_ext(*args, **kwargs)
 
     def test_build_ext(self):
-        cmd = support.missing_compiler_executable()
+        missing_compiler_executable()
         copy_xxmodule_c(self.tmp_dir)
         xx_c = os.path.join(self.tmp_dir, 'xxmodule.c')
         xx_ext = Extension('xx', [xx_c])
@@ -359,7 +360,7 @@ class TestBuildExt(TempdirManager):
         assert cmd.compiler == 'unix'
 
     def test_get_outputs(self):
-        cmd = support.missing_compiler_executable()
+        missing_compiler_executable()
         tmp_dir = self.mkdtemp()
         c_file = os.path.join(tmp_dir, 'foo.c')
         self.write_file(c_file, 'void PyInit_foo(void) {}\n')

--- a/distutils/tests/test_config_cmd.py
+++ b/distutils/tests/test_config_cmd.py
@@ -1,12 +1,11 @@
 """Tests for distutils.command.config."""
 import os
 import sys
-from test.support import missing_compiler_executable
 
 import pytest
 
 from distutils.command.config import dump_file, config
-from distutils.tests import support
+from distutils.tests import support, missing_compiler_executable
 from distutils._log import log
 
 

--- a/distutils/tests/test_install.py
+++ b/distutils/tests/test_install.py
@@ -17,8 +17,7 @@ from distutils.core import Distribution
 from distutils.errors import DistutilsOptionError
 from distutils.extension import Extension
 
-from distutils.tests import support
-from test import support as test_support
+from distutils.tests import support, missing_compiler_executable
 
 
 def _make_ext_name(modname):
@@ -213,7 +212,7 @@ class TestInstall(
         assert found == expected
 
     def test_record_extensions(self):
-        cmd = test_support.missing_compiler_executable()
+        cmd = missing_compiler_executable()
         if cmd is not None:
             pytest.skip('The %r command is not found' % cmd)
         install_dir = self.mkdtemp()

--- a/setup.cfg
+++ b/setup.cfg
@@ -20,8 +20,6 @@ install_requires =
 
 [options.packages.find]
 exclude =
-	build*
-	dist*
 	docs*
 	tests*
 

